device/transport/amqp/devdoc/amqp_twin_client_requirements.md

```diff
@@ -3,106 +3,139 @@
+**SRS_NODE_DEVICE_AMQP_TWIN_16_032: [** The `disableTwinDesiredPropertiesUpdates` method shall send an `AmqpMessage` using the `SenderLink.send` method with the following annotations and properties:
+- `operation` annotation set to `DELETE`.
+- `resource` annotation set to `/notifications/twin/properties/desired`
+- `correlationId` property set to a random integer
+- `body` set to `undefined`. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_16_003: [** The `attach` method shall call its `callback` immediately if the links are already attached. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_033: [** If the `SenderLink.send` call fails the `disableTwinDesiredPropertiesUpdates` method shall call its callback with the error that caused the failure. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_16_007: [** The `attach` method shall call the `getDeviceCredentials` method on the `authenticationProvider` object passed as an argument to the constructor to retrieve the device id. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_034: [** The `disableTwinDesiredPropertiesUpdates` method shall monitor `Message` objects on the `ReceiverLink.on('message')` handler and until a message with the same `correlationId` as the one that was sent is received. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_16_008: [** The `attach` method shall call its callback with an error if the call to `getDeviceCredentials` fails with an error. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_035: [** The `disableTwinDesiredPropertiesUpdates` method shall call its callback with no argument when a response is received **]**
+// TODO: we need to check error codes!
 
-### detach(callback)
+### detach(callback: (err?: Error) => void): void;
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_004: [** The `detach` method shall call its `callback` immediately if the links are already detached. **]**
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_005: [** The `detach` method shall detach the links and call its `callback` with no arguments if the links are successfully detached. **]**
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_006: [** The `detach` method shall call its `callback` with an `Error` if detaching either of the links fail. **]**
+
+### Links
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_007: [** The endpoint argument for attacheReceiverLink shall be `/devices/<deviceId>/twin`. **]**
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_008: [** The link options argument for attachReceiverLink shall be:
+ attach: {
+        properties: {
+          'com.microsoft:channel-correlation-id' : 'twin:<correlationId>',
+          'com.microsoft:api-version' : endpoint.apiVersion
+        },
+        sndSettleMode: 1,
+        rcvSettleMode: 0
+      } **]**
+
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_009: [** The endpoint argument for attacheSenderLink shall be `/device/<deviceId>/twin`. **]**
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_010: [** The link options argument for attachSenderLink shall be:
+ attach: {
+        properties: {
+          'com.microsoft:channel-correlation-id' : 'twin:<correlationId>',
```

Comment:
>twin:<correlationId> [](start = 52, length = 20)

I assume the correlationId needs to be the same between the senderLink and the receiverLink, right?  If so, do we need to specify this requirement?

Reply:
we do test for this. but no requirement. I'll add one.

---
In reply to: [171685162](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171685162) [](ancestors = 171685162)